### PR TITLE
Profile dataloader initialization in FitLoop and update outdated docs

### LINKED ISSUE #21771: Profiler misses dataloader worker initialization (and outdated docs)
### Description & Motivation

The PyTorch Lightning profiler currently fails to capture the most expensive part of data loading: **dataloader worker initialization and prefetching**. 

While the time spent fetching *subsequent* batches during the epoch is correctly captured under the `[TrainingEpochLoop].train_dataloader_next` action, the massive latency spike at the very start of training goes completely unrecorded. Because of this, users experiencing data-loading bottlenecks during startup are forced to monkey-patch private PyTorch Lightning internals (like `FitLoop.setup_data` and `_PrefetchDataFetcher.__iter__`) just to measure their initialization latency.

Here is why this happens:
1. In `FitLoop.setup_data()`, Lightning calls `iter(self._data_fetcher)`. This triggers PyTorch's DataLoader to spawn its multiprocessing workers. If it's an IterableDataset, it also triggers `_PrefetchDataFetcher.__iter__` to eagerly prefetch the first batches.
2. But the profiler isn't attached yet! The profiler hooks (`data_fetcher._start_profiler = self._on_before_fetch`) are only attached *later*, inside `TrainingEpochLoop`.
3. As a result, the initialization phase goes completely unprofiled. 
*(Additionally, the [Profiler documentation](https://lightning.ai/docs/pytorch/stable/tuning/profiler_basic.html#advanced-profiler) still explicitly advertises `get_train_batch` as the profiled action, which has been outdated since the 2021 loops refactor).*

### Pitch

We would like to fix this by adding a profiling hook specifically for dataloader initialization, ensuring the startup latency appears in the standard Fit Profiler Report.

**1. Profile dataloader initialization and worker spawning**
In `src/lightning/pytorch/loops/fit_loop.py` inside `setup_data()`:

```python
        self._data_fetcher = _select_data_fetcher(trainer, RunningStage.TRAINING)
        self._data_fetcher.setup(combined_loader)
        
        with trainer.profiler.profile("setup_train_dataloader"):
            iter(self._data_fetcher)  # Creates the iterator, spawns workers, and prefetches
```

**2. Update Documentation**
Update the profiler docs to remove the outdated mentions of `get_train_batch` and replace them with the actual hooks: `[TrainingEpochLoop].train_dataloader_next` and `setup_train_dataloader`.

We can include a unit test using `BoringModel` to assert that `"setup_train_dataloader"` is successfully recorded in `profiler.profiled_actions`. We are happy to submit a PR for this if the Core team agrees with the approach!

### Alternatives

The only alternative currently available to users is to manually monkey-patch private Lightning classes (e.g., overriding `_FitLoop.setup_data` and `_PrefetchDataFetcher.__iter__`), which is fragile and breaks between minor versions.

### Additional context

_No response_

cc @lantiga @tchaton

### PR DESCRIPTION
**Pull request by yuxin00j**
## What does this PR do?

Fixes #21771
This PR addresses the issue where the PyTorch Lightning profiler fails to capture the latency spike during dataloader worker initialization and early prefetching. 

Specifically, this PR:
1. **Adds a profiling hook:** Wraps `iter(self._data_fetcher)` in `_FitLoop.setup_data()` with the action name `"setup_train_dataloader"`. This ensures that worker spawn and initialization times appear in the standard Fit Profiler Report.
2. **Updates documentation:** Removes the outdated reference to `get_train_batch` (which was deprecated during the 2021 loops refactor) in the `AdvancedProfiler` documentation and replaces it with `setup_train_dataloader`.
3. **Adds unit tests:** Includes a new test `test_setup_train_dataloader_profiled_actions` using `BoringModel` to assert that the `"setup_train_dataloader"` action is correctly recorded in `profiler.profiled_actions`.

- [ ] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: tests/tests_pytorch/profilers/test_profiler.py [hunk 1]
```diff
@@ -731,3 +731,14 @@ def test_profiler_invalid_table_kwargs(tmp_path):
         with pytest.raises(KeyError) as exc_info:
             PyTorchProfiler(table_kwargs={key: None}, dirpath=tmp_path, filename="profile")
         assert exc_info.value.args[0].startswith(f"Found invalid table_kwargs key: {key}.")
+
+
+@skip_advanced_profiler_py312
```
Comment (zhixiangli) [inline tests/tests_pytorch/profilers/test_profiler.py]: Why skip?

File: docs/source-pytorch/tuning/profiler_basic.rst [hunk 2]
```diff
@@ -80,7 +80,7 @@ Once the **.fit()** function has completed, you'll see an output like this:
 
     Profiler Report
 
-    Profile stats for: get_train_batch
+    Profile stats for: setup_train_dataloader
             4869394 function calls (4863767 primitive calls) in 18.893 seconds
```
Comment (zhixiangli) [inline docs/source-pytorch/tuning/profiler_basic.rst]: Is stats below accurate?

File: tests/tests_pytorch/profilers/test_profiler.py (hunk 1 above)
Comment (yuxin00j) [reply tests/tests_pytorch/profilers/test_profiler.py]: Removed this.

File: docs/source-pytorch/tuning/profiler_basic.rst (hunk 2 above)
Comment (yuxin00j) [reply docs/source-pytorch/tuning/profiler_basic.rst]: Updated the doc to use run_training_batch as the example here.

File: tests/tests_pytorch/profilers/test_profiler.py [hunk 3]
```diff
@@ -731,3 +731,13 @@ def test_profiler_invalid_table_kwargs(tmp_path):
...
+
+
+def test_setup_train_dataloader_profiled_actions(tmp_path):
+    """Ensure that the 'setup_train_dataloader' action is successfully recorded in the profiler."""
+    profiler = AdvancedProfiler(dirpath=tmp_path, filename="profiler")
+    model = BoringModel()
+    trainer = Trainer(default_root_dir=tmp_path, fast_dev_run=2, profiler=profiler)
+    trainer.fit(model)
+
+    assert "setup_train_dataloader" in profiler.profiled_actions
```
[the remark this answers is not in the document]
Comment (deependujha) [reply tests/tests_pytorch/profilers/test_profiler.py]: for testing with AdvancedProfiler, one way would be to read the profiled file:

```python
    # Read the profiler output file directly, as `profiler.teardown()` clears `profiler.profiled_actions`.
    path = profiler.dirpath / f"fit-{profiler.filename}.txt"
    assert "setup_train_dataloader" in path.read_text()
```

since SimpleProfiler is also good enough to assert behaviour, going ahead with it.

**Review by deependujha — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
